File: backend/app/models/domain/resume.py
from pydantic import BaseModel, EmailStr, Field


class ContactSchema(BaseModel):
    name: str
    email: EmailStr
    phone: str | None = None
    location: str | None = None
    linkedin: str | None = None
    github: str | None = None
    website: str | None = None


class WorkExperienceEntry(BaseModel):
    company: str
    title: str
    duration: str
    location: str | None = None
    bullets: list[str] = Field(default_factory=list)


class EducationEntry(BaseModel):
    institution: str
    degree: str
    field: str | None = None
    graduation_year: str | None = None


class SkillsSchema(BaseModel):
    languages: list[str] = Field(default_factory=list)
    frameworks: list[str] = Field(default_factory=list)
    tools: list[str] = Field(default_factory=list)
    other: list[str] = Field(default_factory=list)


class ProjectEntry(BaseModel):
    name: str
    description: str
    tech_stack: list[str] = Field(default_factory=list)
    url: str | None = None


class ResumeSchema(BaseModel):
    contact: ContactSchema
    summary: str | None = None
    work_experience: list[WorkExperienceEntry] = Field(default_factory=list)
    education: list[EducationEntry] = Field(default_factory=list)
    skills: SkillsSchema = Field(default_factory=SkillsSchema)
    projects: list[ProjectEntry] = Field(default_factory=list)
    achievements: list[str] = Field(default_factory=list)
    certifications: list[str] = Field(default_factory=list)

File: frontend/types/resume.ts
export interface ContactSchema {
  name: string;
  email: string;
  phone: string | null;
  location: string | null;
  linkedin: string | null;
  github: string | null;
  website: string | null;
}

export interface WorkExperienceEntry {
  company: string;
  title: string;
  duration: string;
  location: string | null;
  bullets: string[];
}

export interface EducationEntry {
  institution: string;
  degree: string;
  field: string | null;
  graduation_year: string | null;
}

export interface SkillsSchema {
  languages: string[];
  frameworks: string[];
  tools: string[];
  other: string[];
}

export interface ProjectEntry {
  name: string;
  description: string;
  tech_stack: string[];
  url: string | null;
}

export interface ResumeSchema {
  contact: ContactSchema;
  summary: string | null;
  work_experience: WorkExperienceEntry[];
  education: EducationEntry[];
  skills: SkillsSchema;
  projects: ProjectEntry[];
  achievements: string[];
  certifications: string[];
}

export type Seniority = "junior" | "mid" | "senior" | "lead" | "executive";
export type Tone = "technical" | "managerial" | "hybrid";

export interface JDAnalysisSchema {
  role_title: string;
  must_have_skills: string[];
  nice_to_have_skills: string[];
  keywords: string[];
  seniority: Seniority;
  tone: Tone;
  responsibilities: string[];
}

export interface DiffItem {
  section: string;
  before: unknown;
  after: unknown;
}

export interface PendingEditResponse {
  proposed_resume_json: ResumeSchema;
  diff: DiffItem[];
}

export interface VersionEntry {
  resumeJson: ResumeSchema;
  label: string;
  timestamp: string;
}

export interface PendingEdit {
  proposedResumeJson: ResumeSchema;
  diff: DiffItem[];
}

export type PendingEditKind = "tailor" | "section";

export interface AppState {
  resumeJson: ResumeSchema | null;
  versionStack: VersionEntry[];
  pendingEdit: PendingEdit | null;
  pendingEditKind: PendingEditKind | null;
}

export interface ParseResponse {
  resume_json: ResumeSchema;
}

export interface JdSourceRequest {
  jd_text?: string;
  jd_url?: string;
}

export interface TailorRequest extends JdSourceRequest {
  resume_json: ResumeSchema;
}

export interface EditRequest {
  resume_json: ResumeSchema;
  section: string;
  action: string;
  instruction: string;
  index?: number;
  identifier?: string;
  entry_index?: number;
  bullet_index?: number;
}

export interface PdfRequest {
  resume_json: ResumeSchema;
}

export class APIError extends Error {
  constructor(
    public status: number,
    public body: unknown,
  ) {
    super(`API request failed with status ${status}`);
    this.name = "APIError";
  }
}
